packages/io/__tests__/cor.ts

```diff
@@ -101,4 +101,32 @@ describe('Test CoR', () => {
     expect(mockBefore).toBeCalled()
     expect(mockAfter).toBeCalled()
   })
+
+  it(`should return ok on any middleware`, async () => {
+    const cor = new CoR()
+    const mockFn = jest.fn()
+    const mockFn1 = jest.fn()
+    const mockFn2 = jest.fn()
+
+    cor.use(async (_, next) => {
+      mockFn()
+      await next()
+    })
+    cor.use(async (ctx, next) => {
+      mockFn1()
+      ctx.ok('ok')
+      await next()
+    })
+    cor.use(async (ctx) => {
+      mockFn2()
+      ctx.ok('hello')
+    })
+
+    const result = await cor.dispatch({})
+
+    expect(result).toEqual('ok')
+    expect(mockFn).toBeCalled()
+    expect(mockFn1).toBeCalled()
+    expect(mockFn2).toBeCalled()
```

Comment:
In this case, `cor` has 3 middlewares
1. call `mockFn()` and pass control to the next middleware
2. call `mockFn1()`, call `ctx.ok('ok')`, pass control to the next middleware
3. call `mockFn2()`, call `ctx.ok('hello')`, exit the control flow.

My question is about step 2. In this step `ctx.ok('ok')` is called and the promise returned by `dispatch` is resolved with `'ok'`. The workflow should finish because caller will get its response `'ok'`. Why should the control flow continue? And what would happen when `ctx.ok` is called twice?

It looks like 2 threads are working simultaneously:
1. `request | mockFn() | mockFn1() | respond`
2. `request | mockFn() | mockFn1() | mockFn2() | unknown behavior`

Reply:
Additional calls to resolve() will not have any effect, the code is similar to this.

```ts
promise = new Promise(reslove => {
    reslove(1)
    // do something...
    reslove(2)
})
```
There is a discussion about "twice reslove": https://stackoverflow.com/a/29491617

Reply:
> Additional calls to resolve() will not have any effect, the code is similar to this.
> 
> ```ts
> promise = new Promise(reslove => {
>     reslove(1)
>     // do something...
>     reslove(2)
> })
> ```
> 
> There is a discussion about "twice reslove": [stackoverflow.com/a/29491617](https://stackoverflow.com/a/29491617)

It's safe to call `resolve` twice indeed because a promise can only be fulfilled once. But this is unexpected.

Reply:
To be more explicit, the workflow should finished mandatorily once `ctx.ok` is called, any updates after it is a side effect.

Reply:
So how do we handle the after hook, for example if I want to know how long the request spend

``` ts
cor.use((ctx, next) => {
	const start = new Date();
	await next()
	console.log('spend', new Date() - start)
})

cor.use((ctx, next) => {
	doSomething();
	ctx.ok('ok')
})
```

This code is converted to a promise equivalent to:
``` ts
new Promise(reslove => {
	const start = new Date();
	
	doSomething();
	reslove('ok') // There is no way to handle the after hook if return directly here

	console.log('spend', new Date() - start)
})
```

Reply:
Let's see how `koa` handles a request

1. a request arrives and is caught by `handleRequest` method defined in `koa.callback`, in this method, koa creates a `ctx` with `req` and `res` objects, then pass `ctx` and `middlewares` into `koa.handleRequest(ctx, middlewares)` method, https://github.com/koajs/koa/blob/master/lib/application.js#L161-L164
2. In `koa.handleRequest(ctx, middlewares)` method, `ctx` is passed into `middlewares` for various pre- and post-treatments, https://github.com/koajs/koa/blob/master/lib/application.js#L188
3. the `middlewares` are composed by [koa-compose](https://github.com/koajs/compose/blob/master/index.js), in the `compose`, middlewares are called nested by `await next()`, `next()` in `middleware_i` is calling `middleware_i+1`, https://github.com/koajs/compose/blob/master/index.js#L42
4. Once the most inside middleware returns a resolved promise, all pre-treatments are done, and all post-treatments start from the inside middleware to outside middleware.
5. Each middleware resolves when its rest logic is executed and passes the control to its outside middleware, until all middlewares are resolved. The logic goes to https://github.com/koajs/koa/blob/master/lib/application.js#L188 and `ctx` is passed to `handleResponse(ctx)`, which calls `respond(ctx)`, `respond(ctx)` calls `res.end(ctx.body)` to send `response` out.

```const Koa = require('koa');
const app = new Koa();

// logger

app.use(async (ctx, next) => {
  await next();
  const rt = ctx.response.get('X-Response-Time');
  console.log(`${ctx.method} ${ctx.url} - ${rt}`);
});

// x-response-time

app.use(async (ctx, next) => {
  const start = Date.now();
  await next();
  const ms = Date.now() - start;
  ctx.set('X-Response-Time', `${ms}ms`);
});

// response

app.use(async ctx => {
  ctx.body = 'Hello World';
});

app.listen(3000);
```

is more like

```
const start = Date.now()
ctx.body = 'Hello World'
const ms = Date.now() - start
ctx.set('X-Response-Time', `${ms}ms`)
const rt = ctx.response.get('X-Response-Time')
console.log(`${ctx.method} ${ctx.url} - ${rt}`)
res.end(ctx.body)
```

Reply:
I think the code is fine, just the case is improper. Once `ctx.ok` is called, the control flow should be inverted from inside to outside instead of keeping moving inside